### PR TITLE
Refactor to use an explicit "Patcher" class for state management.

### PR DESCRIPTION
**Pull request by rfk**

### CODE REVIEW

**Comment by rfk** [comment]
@chilts here's a suggestion for an explicit "Patcher" class and associated lower-level API, based on my usage in https://github.com/mozilla/fxa-oauth-server/pull/208.  The high-level API remains unchanged at `patcher.patch(options, callback)` but you can now step through the methods explicitly by doing `new patcher(options)` and calling its methods.

Thoughts?

**Comment by rfk** [comment]
Worth noting, this made the tests slightly more verbose because they now create a full `Patcher` object, along with all the extra state that it needs by default.  It doesn't seem too bad though.

File: test/change-user.js [hunk 1]
```diff
@@ -5,27 +5,32 @@ var path = require('path')
...
-      user     : 'user',
-      password : 'password',
-      database : 'database',
-    },
+      }
+    })
   }
-
-  patcher.changeUser.call(ctx, function(err) {
+  var p = new patcher(options);
```
Comment (chilts) [inline test/change-user.js]: Nit: Can this be `new Patcher()` (so also `var Patcher = require('../')` above)? Just shows it's a constructor more I think.

**Comment by chilts** [comment]
This is lovely. Full stop. As for the tests being more verbose, that's alright with me. Happy to merge at any time. After that, I'll do a 0.6 release to npm.

**Comment by rfk** [comment]
> Can this be new Patcher()

Good call, I'll change it.  jshint yelled at me about this over in the fxa-oauth-server PR :-)

**Comment by rfk** [comment]
Per IRL discussion, I added public high-level `connect()` and `end()` methods, along with a few more tweaks to ensure the internal state of the object tracks the current state of the db.  This has made the fxa-oauth-server patch *much* cleaner.  @chilts I think this is ready for final r?

**Pull request merged**